src/com/google/javascript/jscomp/AdvancedCoverageInstrumentationCallback.java

```diff
@@ -0,0 +1,205 @@
+
+    BigInteger uniqueIdentifier = parameterMapping.getUniqueIdentifier(combinedParam);
+    BigInteger maxInteger = new BigInteger(Integer.toString(Integer.MAX_VALUE));
+
+    if (uniqueIdentifier.compareTo(maxInteger) > 0) {
+      throw new ArithmeticException(
+          "Unique Identifier exceeds value of Integer.MAX_VALUE, could not encode with Base 64 VLQ");
+    }
+
+    StringBuilder sb = new StringBuilder();
+
+    try {
+      Base64VLQ.encode(sb, uniqueIdentifier.intValue());
+    } catch (IOException e) {
+      // If encoding in Base64 VLQ fails, we will use the original identifier as it will still
+      // maintain obfuscation and partial optimization
+      sb.append(uniqueIdentifier.intValue());
+    }
+
+    parameterMapping.addParamMapping(sb.toString(), combinedParam);
+
+    Node inner_prop = IR
+        .getprop(IR.name("module$exports$instrument$code"), IR.string("instrumentCodeInstance"));
+    Node outer_prop = IR.getprop(inner_prop, IR.string(INSTRUMENT_CODE_FUNCTION_NAME));
+    Node functionCall = IR.call(outer_prop, IR.string(sb.toString()), IR.number(node.getLineno()));
+    Node exprNode = IR.exprResult(functionCall);
+
+    return exprNode.useSourceInfoIfMissingFromForTree(node);
+  }
+
+  private void addBlockNode(AdvancedFileInstrumentationData instrumentationData, List<Node> blocks,
+      String functionScope) {
+    for (Node block : blocks) {
+      instrumentationData.addNode(functionScope, block);
+    }
+  }
+
+  /**
+   * A class the maintains a mapping of unique identifiers to parameter values. It also generates
+   * unique identifiers by creating a counter starting form 0 and increments this value when
+   * assigning a new unique identifier.
+   */
+  private static final class ParameterMapping {
+
+    private final List<String> uniqueIdentifier;
+    private final List<String> paramValue;
+    private BigInteger nextUniqueIdentifier;
+
+    ParameterMapping() {
+      nextUniqueIdentifier = new BigInteger("0");
+      uniqueIdentifier = new ArrayList<>();
+      paramValue = new ArrayList<>();
+    }
+
+    public BigInteger getUniqueIdentifier(String param) {
+
+      nextUniqueIdentifier = nextUniqueIdentifier.add(new BigInteger("1"));
+      return nextUniqueIdentifier;
+    }
+
+    public void addParamMapping(String identifier, String param) {
+      uniqueIdentifier.add(identifier);
+      paramValue.add(param);
+    }
+
+  }
+
+  /**
+   * A class that maintains a list of blocks that are to be instrumented and organises these block
+   * nodes by the function name and what file they are enclosed within.
+   */
+  private static final class AdvancedFileInstrumentationData {
+
+    private final String fileName;
+    private final Map<String, List<Node>> functionScopedNodes;
+
+    AdvancedFileInstrumentationData(String fileName) {
+      this.fileName = fileName;
+      this.functionScopedNodes = new HashMap<>();
```

Comment:
Prefer to use LinkedHashMap here, as it has deterministic iteration order

Even if this particular map is never actually iterated over it's still safest to avoid nondeterminism if this code is ever changed